### ISSUE TITLE
Improved Loss Reporting in KerasModel.fit

### ISSUE THREAD

**Issue by rbharath** [deepchem/deepchem#2058]
At present, `KerasModel.fit` only reports the final average loss over the last `log_every_n` global steps. For a large dataset, this number often doesn't provide much information about the actual variations in the loss curve. I propose that we alter the return type of `KerasModel.fit` to return a list of losses, with an average loss for every `log_every_n` global steps. (That is, whenever a loss is logged, it's also included in the return value). This should provide for much more accurate tracking of the actual learning dynamics for complex datasets.

**Comment by peastman** [comment]
What situations do you envision this being useful for?  It does provide various ways of logging the losses.

**Comment by rbharath** [comment]
The main use case is for plotting loss curves I think. I'm currently working on a model training on a large dataset with a noisy loss curve. Looking at the logging statements, I know that I'm getting only a rough sense of the loss (I'm getting the loss returned once per epoch) since each epoch is pretty long.

**Comment by peastman** [comment]
That's odd.  It should be logging it at whatever logging frequency you requested.

https://github.com/deepchem/deepchem/blob/9436f338aea06ae72d3cf2f55f387ffc037817c5/deepchem/models/keras_model.py#L401-L407

**Comment by rbharath** [comment]
Ah my bad! Wasn't being clear in my earlier statement. The logging works just fine, but I'm using matplotlib to plot the loss curve after training finishes. My epochs take a long time so I've set it to long every 30 seconds or so, meaning something like 20 logging statements per epoch for me, but I only get the final returned loss for the epoch from `fit()`. So that means my matplotlib loss curve doesn't really capture the shift in loss over time but is rather a rough sample plot of the last loss logged from each epoch

**Comment by peastman** [comment]
Perhaps make this a configurable option?  By default it returns just the final loss, but you can pass an argument to request the complete training curve.

**Comment by rbharath** [comment]
Ah good suggestion! That should be quite possible to do

**Comment by rbharath** [comment]
This was implemented in #2060
